allennlp/data/dataset_readers/encoder_decoder.py

```diff
@@ -0,0 +1,105 @@
+from typing import Dict
+import logging
+
+from overrides import overrides
+
+import tqdm
+
+from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
+from allennlp.data.dataset import Dataset
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.fields import TextField
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import Token, Tokenizer, WordTokenizer
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+START_SYMBOL = "@@START@@"
+END_SYMBOL = "@@END@@"
+
+@DatasetReader.register("encoder_decoder")
+class EncoderDecoderDatasetReader(DatasetReader):
+    """
+    Read a tsv file containing paired sequences, and create a dataset for an ``EncoderDecoder`` model.
+    Expected format for each line: <source_sequence_string>\t<target_sequence_string>
+
+    Parameters
+    ----------
+    source_tokenizer : ``Tokenizer``, optional (default=``WordTokenizer()``)
+        Tokenizer to use to split the input sequences into words or other kinds of tokens.
+    target_tokenizer : ``Tokenizer``, optional (default = self.source_tokenizer)
+        Tokenizer to use to split the output sequences (during training) into words or other kinds
+        of tokens.
+    source_token_indexers : ``Dict[str, TokenIndexer]``, optional
+                             (default=``{"tokens": SingleIdTokenizer()}``)
+        Indexers used to define input (source side) token representations.
+    target_token_indexers : ``Dict[str, TokenIndexer]``, optional (default = self.source_token_indexers)
+        Indexers used to define output (target side) token representations.
+    """
+    def __init__(self,
+                 source_tokenizer: Tokenizer = None,
+                 target_tokenizer: Tokenizer = None,
+                 source_token_indexers: Dict[str, TokenIndexer] = None,
+                 target_token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        self._source_tokenizer = source_tokenizer or WordTokenizer()
+        self._target_tokenizer = target_tokenizer or self._source_tokenizer
+        self._source_token_indexers = source_token_indexers or {"tokens": SingleIdTokenIndexer()}
+        self._target_token_indexers = target_token_indexers or self._source_token_indexers
+        self._token_indexers = {"source": self._source_token_indexers,
```

Comment:
`ModelTestCase.setUpModel` expects this field to be set: https://github.com/allenai/allennlp/blob/master/allennlp/common/testing/model_test_case.py#L30

Also, I thought it would be nice to keep things consistent across dataset readers.